### ISSUE TITLE
feat: add top-level `skore.evaluate(estimator, X_test, y_test)` for final test-set evaluation

### ISSUE THREAD

**Issue by raotalha71** [probabl-ai/skore#2520]
### The gap

skore already has the excellent `CrossValidationReport` for **during development** (unfitted estimator + full training data + CV).

What is still missing is the **final honest evaluation** after you have trained your model on all training data:

```python
# Current way (works but verbose + requires knowing internal names)
report = EstimatorReport(
    estimator=final_model,
    X_test=X_test,
    y_test=y_test,
    fit=False
)
report.metrics.summarize()

### Describe the solution you'd like

from skore import evaluate

report = evaluate(final_model, X_test, y_test)   
report.metrics.summarize()
# or even auto-rich output on creation



### Describe alternatives you've considered, if relevant

_No response_

### Additional context

_No response_

**Comment by auguste-probabl** [comment]
Interesting proposal, thanks!

**Comment by raotalha71** [comment]
Thanks for the positive feedback! 😊 I'm glad you find the proposal interesting.
If you're open to it, I'd love to go ahead and prototype a simple version (starting with the thin wrapper for positional args + auto-task detection) and open a PR for review. That way, we can iterate from there and make sure it fits the project's style.
What do you say? Any specific preferences or things to avoid? Excited to get started!

**Comment by MarieSacksick** [comment]
I like the idea of making things shorter and easier for data scientists - it's actually the point of the whole project! I'm puzzled by adding a new function though, I wonder if we can't leverage better what we already have.

What about a solution that alleviates the current option instead?  

I quickly wrote this:  

```python
# %%
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split

X, y = load_iris(return_X_y=True, as_frame=True)
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42)

# %%
from sklearn.tree import DecisionTreeClassifier

model = DecisionTreeClassifier(random_state=42)
model.fit(X_train, y_train)

# %%
from skore import EstimatorReport

report = EstimatorReport(
    model,
    X_test=X_test,
    y_test=y_test)
```

and it works. The fit argument is not required actually (the documentation could be clearer about that).  

What do you think? It's already shorter.

**Comment by raotalha71** [comment]
Hey! Thanks so much for the quick reply and for writing out that example , really appreciate it 
You’re totally right , EstimatorReport(model, X_test=X_test, y_test=y_test) is already quite short and clean (and nice catch that fit isn’t needed ,the docs could definitely highlight that better!).
That said, I still think a top-level evaluate() function would be a meaningful improvement for these reasons:
1. True one-liner ergonomics
```python
## Current
report = EstimatorReport(model, X_test=X_test, y_test=y_test)

## With evaluate()
report = evaluate(model, X_test, y_test)   # ← positional, no keywords, one import
``` 
2. Discoverability for real usersMost people (especially in notebooks or quick scripts) think “I just want to evaluate my model”. They won’t naturally discover EstimatorReport in the middle of the reporting module. evaluate is the verb they are looking for.
3. Room for smart defaults without magic
A dedicated function lets us safely add intelligence that would feel out of place in the constructor:

Automatic task detection with _find_ml_task → instantly surface the right metrics/plots (ROC + PR for classification, prediction error + R² for regression)
Optional auto baseline comparison (DummyClassifier/DummyRegressor)
Cleaner default summary tailored to the task
Perfect future home for probabilistic metrics (CRPS, calibration, sharpness) when someone uses skpro models

It’s exactly the same spirit as skore.train_test_split — a high-level convenience on top of the powerful lower-level tools.
I’m happy to keep it super thin at first (just the nicer API) and we can iterate. Would you be open to me opening a PR with a minimal version so you can see it in action and decide?
Looking forward to your thoughts! 😊

**Comment by jeromedockes** [comment]
I don't really see how the suggested `evaluate` differs from `EstimatorReport`. taking a (possibly fitted) model and data and returning a report is exactly what the constructor does, and any of the mentioned "magic" or improvements could go there directly rather than introducing a new name, a shallow function and a second way of doing the same thing

**Comment by glemaitre** [comment]
From other discussions that sparkle elsewhere but related to this topic, I'm wondering if we could have only a single entry point. 

You could consider that in the user perspective the interest is to only evaluate an estimator, only with a single split or with several splits and currently we are just special casing them because of engineering reason. I witnessed this friction while doing a demo for the hub where the facts to special the reports with tabs was not obvious but it translate with the coding interface certainly. 

Maybe having a function evaluate to dispatch and return the right type of class would be nice instead of trying to merge all classes. The drawback is that the user still needs to figure out why they are different types of report. Then, we would need a compare function in addition. 

What I like here with those function is that "evaluate" or "compare" are actions that return reports which is closer to what one would express in plain English as pointed out by @GaelVaroquaux and @raotalha71 as well.

**Comment by raotalha71** [comment]
Thanks everyone for the super thoughtful and quick feedback ,this is exactly why I love open-source! 🙏
@MarieSacksick ,spot on, once you know EstimatorReport(..., X_test=..., y_test=...) it’s already pretty short, and great catch about fit not being needed (definitely worth clarifying in the docs!).
@jeromedockes ; I completely agree, we shouldn’t add a shallow wrapper just for the sake of it.
@glemaitre ; your idea of a single high-level evaluate entry point is exactly what I was hoping for and even better than my original proposal! I love it.
What if we turn skore.evaluate into the main user-facing action that intelligently dispatches under the hood?
Proposed API:

from skore import evaluate
## 1. Final test-set evaluation (fitted model) → returns EstimatorReport
report = evaluate(model, X_test=X_test, y_test=y_test)
## 2. During development / cross-validation (unfitted model + CV) → returns CrossValidationReport
report = evaluate(model, X=X, y=y, cv=5)
## 3. (Future) Compare multiple models
reports = compare([model1, model2], X=X, y=y)

Benefits:
Users only remember one verb (evaluate) instead of choosing between EstimatorReport vs CrossValidationReport.
Matches plain English: “I just want to evaluate my model.”
We keep all the powerful internal classes unchanged (no breaking changes).
Perfect place to add smart defaults safely: auto-task detection (_find_ml_task), task-tailored default summary, optional baseline comparison, and later probabilistic metrics (CRPS, calibration, sharpness for skpro models ,my favourite area because of my probability background).
Same spirit as skore.train_test_split , high-level convenience on top of the solid internals.
I’m 100% happy to start with a very thin dispatcher version first (just the two main cases above) and iterate based on your feedback.
Would this direction work for the team? Any strong preferences on the API, naming, or things I should avoid?

**Comment by GaelVaroquaux** [comment]
Hi every-one,

I really like the thinking of @raotalha71 for multiple reasons:
- As a user, I don't think in terms of "Reports", I think about the task that I want to achieve. I also don't think in terms of classes for an evaluation, as I don't have in mind that an evaluation is stateful. So, I'm all for "evaluate", and "compare": simple terms, simple grammar that describes the intent and not the implementation.
- As a user of the displays (eg the project.summarize(), which I personally love), I don't make a big difference between an evaluation done with a cross-validation or one done from a precomputed train-test split. These are evaluations, and I expect to find them on the same plot (true story, I was confused not seeing them as such)

On the other hand, with regards to the details of the proposal, I have some hesitations:
- The difference between keyword arguments and non-keyword arguments seems a minor detail, as keyword arguments can be called without giving the name if called in the right order
- More importantly, it feels to me that there are two markedly different signatures that we are compounding in one (I had an IRL discussion with @glemaitre where I had not realized this, so I gave confusing feedback). What worries me is that in one mode (after a prior train-test split) the model is already fitted and the data is only test, while in the other the model will be refitted and the data will be split. I think that it is too error-prone to mix the two.

Based on this thinking, I'd like to propose the following functions:

For evaluating a fitted model
```python
evaluate(model, X_test, y_test)
```

For evaluating a model doing the fit, and the split:
```python
evaluate_cv(model, X, y, cv=5)
```
(note the use of "cv", to be consistent with scikit-learn, and facilitate the on-boarding of users)

For comparing models:
```python
compare(model, X_test, y_test)

compare_cv([model1, model2], X, y, cv=5)

compare_reports([report1, report2])
```

Finally, I'd love a function (technically also usable as a context-manager, in an ideal world):
```python
track(project="my_project", auto_put=True) # Note the default value for the project name: most of the time I don't want to think about a name
```
where, when it is called, the different evaluate functions above automatically put their reports in the project.

Such API would be easier to understand from the user standpoint (more imperative, with verbs that express action) and reduce boilerplate.

Thanks @raotalha71 for giving your inputs, they are much appreciated!!

**Comment by glemaitre** [comment]
> What worries me is that in one mode (after a prior train-test split) the model is already fitted and the data is only test

It depends. You can pass an unfitted model and the train set as well.

I don't know if the splitter parameter with some values could allow to make the dispatching. For instance splitter="single" would do a train test split, passing a splitter will make the CV cross validation and splitter=None would only work for passing a fitted model and a test set.

**Comment by glemaitre** [comment]
After some discussion IRL with @jeromedockes I'm thinking that we could have a single `evaluate` that would take model(s) as input and `compare` that would take reports. In short, the cases would be:

A single model with a single train-test split. Here, it is most probably the trickiest case because, one should either pass a `ShuffleSplit` with a single split or we could create a `TrainTestSplit` splitter in order to expose all the parameters related to the splitter in the signature.

```python
evaluate(
    single_model,
    X=X,
    y=y,
    # `TrainTestSplit` is an equivalent to `ShuffleSplit(n_splits=1)`
    splitter=skore.TrainTestSplit(test_size=0.25)
)
```

A fitted model where we only allow to pass the testing set:

```python
evaluate(single_fitted_model, X=X_test, y=y_test, splitter=None)
```

In both cases, we return an `EstimatorReport`.

A single model using a cross-validation splitter:

```python
evaluate(single_model, X=X, y=y, cv=KFold(n_splits=5))
```

It will return a `CrossValidationReport`.

A list of models (in conjunction of the previous `splitter` parameters)

```python
evaluate([model_1, model_2], X=X, y=y, splitter=KFold(n_splits=5))
```

It will return a `ComparisonReport` containing `EstimatorReport` or `CrossValidationReport` depending of the `splitter`.

In case the latest case, does not provide enough flexibility (e.g. needs to compare two models where `X` is different due to feature engineering), then we provide a `compare` function that consume reports only (and not models):

```python
splitter = skore.TrainTestSplit(shuffle=False)
report_1 = evaluate(model_1, X, y, splitter=splitter)
report_2 = evaluate(model_2, X_engineered, y, splitter=splitter)
comparison = compare([report_1, report_2])
```

**Comment by raotalha71** [comment]
Thanks @glemaitre , @GaelVaroquaux , @jeromedockes  and @MarieSacksick  this discussion is gold! 
I completely agree with the direction: one main evaluate entry point that feels natural in plain English is the way to go.
The latest proposal with the splitter= parameter feels very clean and powerful:

```Python
 #1. Fitted model + test set only
report = evaluate(model, X=X_test, y=y_test, splitter=None)

#2. Unfitted model + train-test split
report = evaluate(model, X=X, y=y, splitter=TrainTestSplit(test_size=0.25))

#3. Unfitted model + cross-validation
report = evaluate(model, X=X, y=y, splitter=KFold(n_splits=5))

#4. Multiple models
comparison = evaluate([model1, model2], X=X, y=y, splitter=KFold(n_splits=5))
```
This solves discoverability, reduces user confusion, and gives us a perfect place to add smart defaults later (auto-task detection, baseline comparison, probabilistic metrics from skpro, etc.).
I’d love to take this forward by implementing a working prototype of the dispatcher (starting with the three main cases above) and open a PR so we can test it live and iterate quickly.
Any preferences on:
Whether to start with splitter= or the separate evaluate / evaluate_cv version?
Should I also include the compare function in the first PR or keep it minimal?
Happy to follow whatever direction the team prefers ,just let me know and I’ll start coding today.
Thanks again for the thoughtful input , excited to help make this happen!

**Comment by jeromedockes** [comment]
It's a bit of a detail but I think we need something less subtle than splitter=None to mean "no splitting", as cv=None in scikit-learn mans "default splitting strategy" so users may not pay attention to it and it ends up doing something radically different. maybe we can prevent such mistakes by refusing to fit an already-fitted estimator passed together with a splitter, and refusing to make predictions with an unfitted estimator (which makes no sense for most estimators anyway) when no splitting is done

**Comment by GaelVaroquaux** [comment]
> It's a bit of a detail but I think we need something less subtle than splitter= None to mean "no splitting", as cv=None in scikit-learn mans "default splitting strategy" so users may not pay attention to it and it ends up doing something radically different.

I agree

> maybe we can prevent such mistakes by refusing to fit an already-fitted estimator passed together with a splitter, and refusing to make predictions with an unfitted estimator (which makes no sense for most estimators anyway) when no splitting is done

Seems like a good idea to me, alongside with really good error messages

**Comment by GaetandeCast** [comment]
A few things we thought about when discussing this IRL with @glemaitre:
- It would be nice to be able to pass a float to `splitter` to implicitly call `sklearn.train_test_split()` with default parameters and `test_size` equal to that float. Then the default value of `splitter` could be `0.2`. This lets us train and evaluate a model with the scikit-learn defaults in a natural way.
- For comparisons, we need to let `X` accept a list of datasets too, as we can compare estimators that have different input features. We then need to check that there is either a common `X` for all estimators or as many estimators as datasets. Also, for now at least, `y` should be the same for all estimators as we don't allow comparing estimators with different targets in skore yet. `splitter` could also be either a list or a single element.

**Comment by raotalha71** [comment]
I'll implement evaluate() with the following spec based on the discussion:
```python 
evaluate(estimator, X, y, splitter=0.2)

splitter=float → train_test_split(test_size=float) → returns EstimatorReport
splitter=CVObject → cross-validation → returns CrossValidationReport
splitter=None → expects already-fitted estimator + test data only → returns EstimatorReport
estimator=list → returns ComparisonReport```

Validation rules:

fitted model + splitter given → raise error
unfitted model + splitter=None → raise error
clear, actionable error messages in both cases
```
I'll keep the first PR minimal — just the dispatcher + validation + tests. No extra magic until this is reviewed.

**Comment by raotalha71** [comment]
Thanks everyone for the incredibly productive discussion! Before I start coding the PR, I want to make sure I have the spec right. Here's my understanding + a few open questions:

## My understanding of the agreed API

```python
from skore import evaluate

# 1. Unfitted model + train-test split (default) → EstimatorReport
report = evaluate(model, X, y)                          # splitter=0.2 by default
report = evaluate(model, X, y, splitter=0.3)            # custom test_size

# 2. Fitted model + test set only → EstimatorReport
report = evaluate(fitted_model, X_test, y_test, splitter=???)  # see Q1 below

# 3. Unfitted model + cross-validation → CrossValidationReport
report = evaluate(model, X, y, splitter=KFold(n_splits=5))

# 4. Multiple models → ComparisonReport
report = evaluate([model1, model2], X, y, splitter=KFold(n_splits=5))
```

Validation guards (per @jeromedockes ):

Fitted estimator + splitter provided → raise with clear error message
Unfitted estimator + no splitting → raise with clear error message
Open questions
Q1: How to signal "no splitting" for a pre-fitted model?
@jeromedockes  raised that splitter=None is dangerous because cv=None in scikit-learn means "use default CV strategy" — users might pass None expecting default behavior and get something completely different.
Options I see:
(a) splitter=None but rely on the fitted-vs-unfitted validation to catch mistakes
(b) A string sentinel like splitter="prefit" (explicit and hard to confuse)
(c) Something else?

Q2: Should splitter=int (e.g. splitter=5) mean KFold(n_splits=5)?
CrossValidationReport already accepts splitter=int and interprets it as the number of CV folds. Should evaluate() do the same? This would give us:
```
evaluate(model, X, y, splitter=5)  # → CrossValidationReport with 5 folds
```
Q3: Should the float-based split use skore's train_test_split (with the smart warnings) or plain sklearn's?
Using skore's version would surface the data leakage / class imbalance / temporal warnings automatically, which seems aligned with the project's philosophy. But I want to confirm.

What I'll include in the first PR
The evaluate() dispatcher function (single new file)
Wire the export in __init__.py
Tests for all 4 cases + validation error cases
No compare() function (separate PR)
No track() / auto-project integration (future)
No extra magic (auto-baseline, probabilistic metrics — future PRs)
Looking forward to your input on Q1-Q3 so I can start coding with confidence!

**Comment by jeromedockes** [comment]
my 2cents on the questions:

Q1 I would go for your option b, a special string to mean no splitting
Q2 yes
Q3 I guess skore's

**Comment by raotalha71** [comment]
Hi @glemaitre  @jeromedockes , I've opened a PR for this: #2573. It implements the evaluate() dispatcher based on the design we discussed here. Would appreciate your feedback when you get a chance!

**Comment by raotalha71** [comment]
Thanks Everyone ! means alot.
